redis/src/main/scala/zio/redis/api/PubSub.scala

```diff
@@ -0,0 +1,89 @@
+package zio.redis.api
+
+import zio.redis.Input._
+import zio.redis.Output._
+import zio.redis.ResultBuilder.ResultOutputStreamBuilder
+import zio.redis._
+import zio.redis.api.PubSub._
+import zio.schema.Schema
+import zio.stream._
+import zio.{Chunk, ZIO}
+
+trait PubSub {
+  final def subscribeStreamBuilder(channel: String, channels: String*): ResultOutputStreamBuilder =
+    new ResultOutputStreamBuilder {
+      override def returning[R: Schema]: ZStream[Redis, RedisError, R] =
+        ZStream.serviceWithStream[Redis] { redis =>
+          RedisPubSubCommand
+            .run(RedisPubSubCommand.Subscribe(channel, channels.toList))
+            .collect { case t: PushProtocol.Message => t.message }
+            .mapZIO(resp =>
+              ZIO
+                .attempt(ArbitraryOutput[R]().unsafeDecode(resp)(redis.codec))
+                .refineToOrDie[RedisError]
+            )
+        }
+    }
+
+  final def subscribe(channel: String, channels: String*): ZStream[Redis, RedisError, PushProtocol] =
```

Comment:
> In case of multiple channels passed, we should return a List/Chunk of streams.

okay then it'd be better to get api with a single param and the callers have to handle multiple params using `ZIO.foreach`

> I think that we should hide PushProtocol from the end user.

hmm... meaning that defines separate data types containing pushed message content for only end users or registers callbacks for each pubsub command for the users who want to handle subscribe/unsubscribe messages?

Reply:
or using ResultBuilder that ask to give callbacks of sub/unsub when calls `returning`

Reply:
```scala
def subscribe(channel: String): ResultOutputStreamBuilder =
  new ResultOutputStreamBuilder(
    override def returning[R: Schema]: ZStream[Redis, RedisError, R] =
        ZStream.serviceWithStream[Redis] { redis =>
          RedisPubSubCommand
            .run(RedisPubSubCommand.Subscribe(channel))
            .collect { case t: PushProtocol.Message => t.message }
            .mapZIO { resp =>
              ZIO
                .attempt(ArbitraryOutput[R]().unsafeDecode(resp)(redis.codec))
                .refineToOrDie[RedisError]
            }
        }
  )

def subscribe(channels: Chunk[String]): ResultOutputMultiStreamBuilder =
  new ResultOutputMultiStreamBuilder(
    override def returning[R: Schema]: UIO[Chunk[ZStream[Redis, RedisError, R]]] =
        ZIO.serviceWithZIO[Redis] { redis =>
          RedisPubSubCommand
            .run(RedisPubSubCommand.Subscribe(channels.head, channels.tail))
            .collect { case t: PushProtocol.Message => t.message }
            .broadcast(channels.length, maximumLag)
            .map(streams =>
              streams.zipWith(channels) { (stream, channel) =>
                stream
                  .filter(_.key.value == channel)
                  .mapZIO { resp =>
                    ZIO
                      .attempt(ArbitraryOutput[R]().unsafeDecode(resp)(redis.codec))
                      .refineToOrDie[RedisError]
                  }
              }
            )
        }
```
This is just a sketch. I hope that I managed to explain what's the base idea through this code. Of course, something like `ResultOutputMultiStreamBuilder` should be added for multi-stream scenarios.

Reply:
yeah that's almost same what I though and mentioned above 👍